ddprof-ffi/src/tags.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{AsBytes, CharSlice};
+use ddprof_exporter::parse_tags;
+use ddprof_exporter::tag::Tag;
+use std::error::Error;
+
+#[must_use]
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_new() -> crate::Vec<Tag> {
+    crate::Vec::default()
+}
+
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_drop(_: crate::Vec<Tag>) {}
+
+#[repr(C)]
+pub enum PushTagResult {
+    Ok,
+    Err(crate::Vec<u8>),
+}
+
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_PushTagResult_drop(_: PushTagResult) {}
+
+/// Creates a new Tag from the provided `key` and `value` by doing a utf8
+/// lossy conversion, and pushes into the `vec`. The strings `key` and `value`
+/// are cloned to avoid FFI lifetime issues.
+///
+/// # Safety
+/// The `vec` must be a valid reference.
+/// The CharSlices `key` and `value` must point to at least many bytes as their
+/// `.len` properties claim.
+#[must_use]
+#[no_mangle]
+pub unsafe extern "C" fn ddprof_ffi_Vec_tag_push(
+    vec: &mut crate::Vec<Tag>,
+    key: CharSlice,
+    value: CharSlice,
+) -> PushTagResult {
+    let key = key.to_utf8_lossy().into_owned();
+    let value = value.to_utf8_lossy().into_owned();
+    match Tag::new(key, value) {
+        Ok(tag) => {
+            vec.push(tag);
+            PushTagResult::Ok
+        }
+        Err(err) => PushTagResult::Err(err.as_bytes().to_vec().into()),
+    }
+}
+
+#[repr(C)]
+pub struct ParseTagsResult {
+    tags: crate::Vec<Tag>,
+    error_message: Option<Box<crate::Vec<u8>>>,
+}
+
+#[must_use]
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_parse(string: CharSlice) -> ParseTagsResult {
+    match unsafe { string.try_to_utf8() } {
+        Ok(string) => {
+            let (tags, error) = parse_tags(string);
+            ParseTagsResult {
+                tags: tags.into(),
+                error_message: error
+                    .map(|message| Box::new(crate::Vec::from(message.into_bytes()))),
+            }
+        }
+        Err(err) => ParseTagsResult {
+            tags: crate::Vec::default(),
+            error_message: Some(Box::new(crate::Vec::from(&err as &dyn Error))),
+        },
+    }
+}
+
+#[must_use]
+#[allow(clippy::ptr_arg)]
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_Vec_tag_clone(tags: &crate::Vec<Tag>) -> crate::Vec<Tag> {
+    tags.iter()
+        .map(|tag| tag.clone().into_owned())
+        .collect::<Vec<Tag>>()
+        .into()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::tags::*;
+
+    #[test]
+    fn empty_tag_name() {
+        unsafe {
+            let mut tags = ddprof_ffi_Vec_tag_new();
+            let result =
+                ddprof_ffi_Vec_tag_push(&mut tags, CharSlice::from(""), CharSlice::from("woof"));
+            assert!(!matches!(result, PushTagResult::Ok));
```

Comment:
Hmm... wasn't a tag with an empty key invalid?

Reply:
Note the negation: `assert!(!matches!(result, PushTagResult::Ok))`. It's _not_ ok. It's done this way because `Err` has associated data, but Ok does not in this case.

Reply:
!there's)!too!(many!exclamation!marks!in!a)!handful!((of!characters!))))lol!thanks!for!clearing!it!up